sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java

```diff
@@ -88,108 +100,220 @@ public static void main(String[] args) throws IOException, InterruptedException
+        for (Map.Entry<String, String> twin : twins.entrySet()) {
+            String twinId = twin.getKey();
+
+            List<BasicRelationship> relationshipList = new ArrayList<>();
+            Semaphore listRelationshipSemaphore = new Semaphore(0);
+            Semaphore deleteRelationshipsSemaphore = new Semaphore(0);
+            CountDownLatch deleteTwinsLatch = new CountDownLatch(1);
+
+            // Call APIs to retrieve all relationships.
+            client.listRelationships(twinId, BasicRelationship.class)
+                .doOnNext(relationshipList::add)
+                .doOnError(IgnoreNotFoundError)
+                .doOnTerminate(listRelationshipSemaphore::release)
+                .subscribe();
+
+            // Call APIs to retrieve all incoming relationships.
+            client.listIncomingRelationships(twinId)
+                .doOnNext(e -> relationshipList.add(mapper.convertValue(e, BasicRelationship.class)))
+                .doOnError(IgnoreNotFoundError)
+                .doOnTerminate(listRelationshipSemaphore::release)
+                .subscribe();
+
+            // Call APIs to delete all relationships.
+            if (listRelationshipSemaphore.tryAcquire(2, MaxWaitTimeAsyncOperationsInSeconds, TimeUnit.SECONDS)) {
+                relationshipList
+                    .forEach(relationship -> client.deleteRelationship(relationship.getSourceId(), relationship.getId())
+                        .doOnSuccess(aVoid -> {
+                            if (twinId.equals(relationship.getSourceId())) {
+                                System.out.println("Found and deleted relationship: " + relationship.getId());
+                            } else {
+                                System.out.println("Found and deleted incoming relationship: " + relationship.getId());
+                            }
+                        })
+                        .doOnError(IgnoreNotFoundError)
+                        .doOnTerminate(deleteRelationshipsSemaphore::release)
+                        .subscribe());
+            }
+
+            // Verify that the relationships have been deleted.
+            if (deleteRelationshipsSemaphore.tryAcquire(relationshipList.size(), MaxWaitTimeAsyncOperationsInSeconds, TimeUnit.SECONDS)) {
+                // Now the digital twin should be safe to delete
+
+                // Call APIs to delete the twins.
+                client.deleteDigitalTwin(twinId)
+                    .doOnSuccess(aVoid -> System.out.println("Deleted digital twin: " + twinId))
+                    .doOnError(IgnoreNotFoundError)
+                    .doOnTerminate(deleteTwinsLatch::countDown)
+                    .subscribe();
+
+                // Wait until the latch has been counted down for each async delete operation, signifying that the async call has completed successfully.
```

Comment:
For each async delete? Can you elaborate what you mean here?

Reply:
What is a latch?

Reply:
Appears to be some kind of threadsafe counter, eh?

I'd reword this to be "Wait until the latch count reaches zero..."

Reply:
A countdown latch is a mechanism to block the calling thread until other threads that are running in parallel have completed (counted down). The `.countdown()` essentially decrements a thread-safe counter.

From this article online: https://www.baeldung.com/java-countdown-latch
Simply put, a CountDownLatch has a counter field, which you can decrement as we require. We can then use it to block a calling thread until it's been counted down to zero.

If we were doing some parallel processing, we could instantiate the CountDownLatch with the same value for the counter as a number of threads we want to work across. Then, we could just call countdown() after each thread finishes, guaranteeing that a dependent thread calling await() will block until the worker threads are finished.